README.md
**txs** (**t**est **x**264 **s**ettings) is a command line tool that can
generate different combinations of x264 settings and make short test encodings
with [ffmpeg](https://ffmpeg.org). It also comes with a Lua script for
[mpv](https://mpv.io) that allows you to visually compare the test encodes and
find the best one based on quality, estimated file size and estimated encoding
time. It looks like this:

![Demo](https://github.com/plotski/txs/blob/master/demo.gif?raw=true)

This is the output of `txs tutorial`:

This is a tutorial that should get you started. Run `txs -h` and
`txs SUBCOMMAND -h` for more information.

    $ txs -s source.mkv -r 25:00 10 -x crf=19:me=umh samples \
      -xs subme=9/10:deblock=-2,-1/-3,-3:no-fast-pskip

The above command creates test encodes or samples with all possible combinations
of the given values for "subme", "deblock" and "no-fast-pskip" in the directory
"samples.orig@25:00-1.subme:deblock:no-fast-pskip". The samples are all encoded
with "crf=19:me=umh", they are all 10 seconds long and start at 25 minutes in
source.mkv:

    source.sample@25:00-10.crf=19:me=umh:subme=10:deblock=-2,-1.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10:deblock=-2,-1:no-fast-pskip.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10:deblock=-3,-3.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10:deblock=-3,-3:no-fast-pskip.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=9:deblock=-2,-1.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=9:deblock=-2,-1:no-fast-pskip.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=9:deblock=-3,-3.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=9:deblock=-3,-3:no-fast-pskip.mkv

By providing multiple sets of sample settings to -xs you can limit the number of
combinations:

    $ txs -s source.mkv -r 25:00 10 -x crf=19:me=umh samples \
      -xs subme=10/11:no-deblock deblock=-2,-1/-3,-3

This creates the following samples:

    source.sample@25:00-10.crf=19:me=umh:deblock=-2,-1.mkv
    source.sample@25:00-10.crf=19:me=umh:deblock=-3,-3.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10:no-deblock.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=11.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=11:no-deblock.mkv

To compare encodes, use the "compare" subcommand:

    $ txs -s source.mkv compare samples.orig@25:00-1.subme:deblock:no-fast-pskip

This opens mpv in fullscreen mode with a playlist of two samples. Switch between
samples with "j" and "k" or toggle the original source with "o". Pick the better
sample with "b", the worse with "w" or mark them as equal with "e". New samples
are loaded automatically every time you make a decision. After you've rated all
samples once, any samples that were marked as equal are loaded again. This
process will leave you with the best sample and its settings in the end.

You can adjust the playlist size with the -p (--playlist-size) option if you
want to compare more than two samples.

"Shift+w" does the same thing as "w", but it also removes the sample and its log
file from the file system.

Show and hide the playlist overlay with "`".

Some final notes you might find useful:

- Choose your sample range carefully. It should be representative of the full
  video. 10 seconds or less is fine to check something quickly, but use 60
  seconds or more for fine tuning.

- Increasing gamma with "6" to around 10 to 15 makes differences more obvious.
  (Decrease with "5".) Don't go too high or you'll watch pixels dance that
  nobody else will ever see.

- Try switching between samples while playing. txs should preserve playback
  time, but there is a small delay when mpv seeks.

- You can seek forward and backward by single frames with "." and ",".

### Installation

Install [pipx](https://pipxproject.github.io/pipx/) with your distro's package
manager or with pip:

    $ python3 -m pip install --user pipx

Then install txs with pipx:

    $ pipx install txs

Upgrade:

    $ pipx upgrade txs

Install development version over current release:

    $ pipx upgrade --spec git+https://github.com/plotski/txs.git txs

Install development version without an existing txs installation:

    $ pipx install --spec git+https://github.com/plotski/txs.git txs

Uninstall:

    $ pipx uninstall pipx


NEWS
2020-06-11 0.1.3
  * Fix previous fix


2020-06-11 0.1.2
  * Fix previous release


2020-06-11 0.1.1
  * Fix potential installation issue due to importing txs during installation


2020-02-16 0.1.0
  * Fix mpv error "Error parsing commandline option sub-file: option requires
    parameter" due to mpv 0.32 demanding arguments as "--option=parameter"
  * Add support for BluRay directories as source: directories with a "BDMV"
    subdirectory are passed as "bluray:<path>" to ffmpeg


2019-12-30 0.0.3
  * Fix syntax error in setup.py
  * Print help if no arguments are given


2019-12-30 0.0.2
  * Copy audio in test encodes get more interesting final size estimates
  * Attempt to make it work on Windows and Mac (untested)
  * Bugfix: Delete test encode log file when Ctrl-c is pressed
  * Bugfix: Ensure order of settings so the txs-compare.lua can find estimates


2019-12-23 0.0.1
  * Initial release


txs/utils.py
import sys
import os
import site
import re
import itertools
from collections import abc, defaultdict
import shlex
import subprocess
import textwrap
import termios, tty
import contextlib

from . import utils
from . import __name__

def error(msg):
    print(msg, file=sys.stderr)

def croak(msg=None):
    if msg:
        error(msg)
    sys.exit(1)

@contextlib.contextmanager
def raw_mode_posix():
    attrs = termios.tcgetattr(sys.stdin.fileno())
    tty.setraw(sys.stdin.fileno())
    try:
        yield
    finally:
        termios.tcsetattr(sys.stdin.fileno(), termios.TCSADRAIN, attrs)

def dialog_yesno(question):
    answer = ''
    try:
        if os.name == 'posix':
            print(f'{question} [y/n] ', end='', flush=True)
            with raw_mode_posix():
                while answer.lower() not in ('y', 'n'):
                    answer = sys.stdin.read(1)
            print(answer)
        else:
            while answer.lower() not in ('y', 'n'):
                answer = input(f'{question} [y/n] ')
    except KeyboardInterrupt:
        pass
    return answer.lower() == 'y'

def wrap(string, width=80):
    return textwrap.fill(string, width=width)

def indent(amount, string):
    return textwrap.indent(string, ' '*amount)

def cmd2str(cmd):
    return ' '.join(shlex.quote(arg) for arg in cmd)

def duration2str(seconds):
    hours = int(seconds / 3600)
    mins = int((seconds - (hours * 3600)) / 60)
    return f'{hours:02d}:{mins:02d}'

def bytes2str(bytes):
    for size,unit in ((2**30, 'Gi'), (2**20, 'Mi'), (2**10, 'Ki')):
        if bytes >= size:
            return f'{round(bytes / size, 2):5.2f} {unit}B'
    return f'{bytes} B'

def combine_dicts(*dcts):
    # {**a, **b} doesn't cut it because the order of the keys in later dicts
    # trumps the order of the same keys in earlier dicts.
    # >>> a = {'me': 'umh', 'bframes': 5}
    # >>> b = {'crf': 20, 'bframes': 5}
    # >>> {**a, **b}
    # {'me': 'umh', 'bframes': 5, 'crf': 20}
    # 'crf' and 'bframes' were flipped, bu we need to maintain key order of b or
    # the Lua script won't find the estimates.
    combined = {}
    for d in dcts:
        for k in d:
            combined.pop(k, None)
        combined.update(d)
    return combined

def logfile(filepath):
    return os.path.splitext(filepath)[0] + '.log'

def title(filepath):
    return os.path.splitext(os.path.basename(filepath))[0]

def mkdir(path):
    if not os.path.exists(path):
        try:
            os.mkdir(path)
        except OSError as e:
            croak(f'Unable to create {path}: {os.strerror(e.errno)}')
    if not os.path.isdir(path):
        croak(f'Not a directory: {path}')

def cleanup(*filepaths):
    for filepath in filepaths:
        for f in (filepath, logfile(filepath)):
            if os.path.exists(f):
                print(f'Deleting {f}')
                os.remove(f)


def parse_settings(*strings, default_value=None):
    # Split settings at unescaped ":"
    strings = [s.replace('\\:', ':') for string in strings
               for s in re.split(r'(?<!\\):', string)]
    # Convert list of "k=v" strings into dict
    settings = {}
    for string in strings:
        if '=' in string:
            k, v = string.split('=', maxsplit=1)
            settings[k.strip()] = v.strip()
        else:
            key = string.strip()
            if key:
                settings[key] = default_value
    return settings

def generate_sample_settings(*strings):
    # Parse each string as a list of settings
    combinations = []
    for string in strings:
        # Each setting can have multiple values separated by unescaped "/"
        settings = parse_settings(string)
        for k,v in settings.items():
            if v is not None:
                settings[k] = [v.replace('\\/', '/')
                               for v in re.split(r'(?<!\\)/', v)]
            else:
                # Flags (e.g. no-deblock) have no value
                settings[k] = [None]

        # Create all combinations of all settings
        # https://stackoverflow.com/a/5228294
        keys = settings.keys()
        value_lists = settings.values()
        for values in itertools.product(*value_lists):
            d = dict(zip(keys, values))
            # For each encountered flag (boolean setting), add another sample
            # with that flag removed.
            for k,v in tuple(d.items()):
                if v is None:
                    d2 = d.copy()
                    d2.pop(k)
                    combinations.append(d2)
            combinations.append(d)
    return combinations

def sample_keys(sample_settings):
    # Same thing as set().union(), but preserve order.
    keys = []
    for settings in sample_settings:
        for key in settings.keys():
            if key not in keys:
                keys.append(key)
    return keys

def settings2str(settings, delimiter=':', escape=False, replace_in_values={':':','}):
    # Values that use ":" as a separator (e.g. deblock) can also use ",", which
    # makes the whole string easier to parse.
    def apply_riv(value):
        for this,that in replace_in_values.items():
            value = str(value).replace(this, that)
        return value

    def normalize_value(value):
        if value is None:
            return None
        value = apply_riv(value)
        if escape:
            value = value.replace(delimiter, '\\'+delimiter)
            value = value.replace('=', '\\=')
        return value

    if isinstance(settings, abc.Sequence):
        # Group together settings with identical key sets and map each group to
        # a tuple of the shared keys.
        # Example: {('crf', 'bframes'): [{'crf': '22', 'bframes': '8'},
        #                                {'crf': '22', 'bframes': '16'},
        #                                {'crf': '21', 'bframes': '8'},
        #                                {'crf': '21', 'bframes': '16'}],
        #           ('b-adapt', 'bframes'): [{'b-adapt': '1', 'bframes': '3'},
        #                                    {'b-adapt': '1', 'bframes': '6'},
        #                                    {'b-adapt': '2', 'bframes': '3'},
        #                                    {'b-adapt': '2', 'bframes': '6'}]}
        groups = defaultdict(lambda: [])
        for settings in settings:
            groups[tuple(settings.keys())].append(settings)
        # For each group, map each key to a list of values.
        strings = []
        for keys,list_of_settings in groups.items():
            value_lists = defaultdict(lambda: [])
            for key in keys:
                value_list = value_lists[key]
                for settings in list_of_settings:
                    value = settings[key]
                    if value not in value_list:
                        value = normalize_value(value)
                        if value:
                            value_list.append(value)
            # Separate value with "/"
            parts = []
            for key,values in value_lists.items():
                if len(values) > 0:
                    parts.append(f'{key}=' + '/'.join(values))
                else:
                    # key is a flag/boolean
                    parts.append(key)
            # Separate settings with ":"
            strings.append(delimiter.join(parts))
        # Separate groups of settings with " "
        return ' '.join(strings)

    else:
        parts = []
        for k,v in settings.items():
            value = normalize_value(v)
            if value:
                parts.append(f'{k}={normalize_value(v)}')
            else:
                parts.append(f'{k}')
        return delimiter.join(parts)

def read_estimates(estimates_file):
    est = {}
    if os.path.exists(estimates_file):
        with open(estimates_file, 'r') as f:
            for line in f.readlines():
                parts = [part.strip() for part in line.split('/')]
                est[parts[0]] = {'settings'     : parts[0],
                                 'time_str'     : parts[1],
                                 'time'         : parts[2],
                                 'size_str'     : parts[3],
                                 'size'         : parts[4],
                                 'all_settings' : parts[5]}
    return est

def update_estimates(estimates_file, diff_settings, est_time, est_size, settings):
    est = read_estimates(estimates_file)
    diff_settings_str = utils.settings2str(diff_settings, escape=False)
    est[diff_settings_str] = {'settings'     : diff_settings_str,
                              'time_str'     : utils.duration2str(est_time),
                              'time'         : int(est_time),
                              'size_str'     : utils.bytes2str(est_size),
                              'size'         : int(est_size),
                              'all_settings' : utils.settings2str(settings, escape=True)}
    max_key_width = max(len(k) for k in est)
    with open(estimates_file, 'w') as f:
        for key,values in est.items():
            values = ' / '.join(str(v) for v in tuple(values.values())[1:])
            f.write(f'{key.ljust(max_key_width)} / {values}\n')

if os.name == 'posix':
    MPV = 'mpv'
elif os.name == 'nt':
    MPV = 'mpv.exe'
else:
    raise RuntimeError('Unsupported os: {os.name!r}')

def compare_samples(sample_dir, debug=None, playlist_size=None, font_size=None, estimates_file=None):
    script_path_user = os.path.join(site.USER_BASE, f'share/{__name__}/lua/{__name__}-compare.lua')
    script_path_system = os.path.join(sys.prefix, f'share/{__name__}/lua/{__name__}-compare.lua')
    if os.path.exists(script_path_user):
        script_path = script_path_user
    elif os.path.exists(script_path_system):
        script_path = script_path_system
    else:
        error(f'No such file: {script_path_user}')
        error(f'No such file: {script_path_system}')
        croak(f'Cannot find {__name__}-compare.lua')

    cmd = [MPV, '--idle']
    cmd.append(f'--script={script_path}')
    scriptopts = []
    if debug:
        scriptopts.append(f'{__name__}-debug={"yes" if debug else "no"}')
    if playlist_size:
        scriptopts.append(f'{__name__}-playlist_size={playlist_size}')
    if font_size:
        scriptopts.append(f'{__name__}-font_size={font_size}')
    if estimates_file:
        scriptopts.append(f'{__name__}-estimates_file={estimates_file}')
    if scriptopts:
        cmd.append(f'--script-opts={",".join(scriptopts)}')
    if debug:
        print(cmd2str(cmd))
    subprocess.run(cmd, cwd=sample_dir)


txs/main.py
import os
import time
import argparse
import sys
from collections import abc
from . import utils
from . import ffmpeg
from . import __name__, __version__

class MyHelpFormatter(argparse.HelpFormatter):
    def _get_help_string(self, action):
        def as_str(thing):
            if isinstance(thing, abc.Iterable) and not isinstance(thing, str):
                string = ' '.join(as_str(x) for x in thing)
            elif not isinstance(thing, bool):
                string = str(thing)
            else:
                string = None
            if string:
                return string

        help = action.help
        if action.default is not argparse.SUPPRESS:
            defaulting_nargs = [argparse.OPTIONAL, argparse.ZERO_OR_MORE]
            if action.option_strings or action.nargs in defaulting_nargs:
                string = as_str(action.default)
                if string is not None:
                    help += f' (Default: {string})'
        return help

    def _fill_text(self, text, width, indent):
        return ''.join(indent + line for line in text.splitlines(keepends=True))


TUTORIAL = f'''
This is a tutorial that should get you started. Run `{__name__} -h` and
`{__name__} SUBCOMMAND -h` for more information.

    $ {__name__} -s source.mkv -r 25:00 10 -x crf=19:me=umh samples \\
      -xs subme=9/10:deblock=-2,-1/-3,-3:no-fast-pskip

The above command creates test encodes or samples with all possible combinations
of the given values for "subme", "deblock" and "no-fast-pskip" in the directory
"samples.orig@25:00-1.subme:deblock:no-fast-pskip". The samples are all encoded
with "crf=19:me=umh", they are all 10 seconds long and start at 25 minutes in
source.mkv:

    source.sample@25:00-10.crf=19:me=umh:subme=10:deblock=-2,-1.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10:deblock=-2,-1:no-fast-pskip.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10:deblock=-3,-3.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10:deblock=-3,-3:no-fast-pskip.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=9:deblock=-2,-1.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=9:deblock=-2,-1:no-fast-pskip.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=9:deblock=-3,-3.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=9:deblock=-3,-3:no-fast-pskip.mkv

By providing multiple sets of sample settings to -xs you can limit the number of
combinations:

    $ {__name__} -s source.mkv -r 25:00 10 -x crf=19:me=umh samples \\
      -xs subme=10/11:no-deblock deblock=-2,-1/-3,-3

This creates the following samples:

    source.sample@25:00-10.crf=19:me=umh:deblock=-2,-1.mkv
    source.sample@25:00-10.crf=19:me=umh:deblock=-3,-3.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=10:no-deblock.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=11.mkv
    source.sample@25:00-10.crf=19:me=umh:subme=11:no-deblock.mkv

To compare encodes, use the "compare" subcommand:

    $ {__name__} -s source.mkv compare samples.orig@25:00-1.subme:deblock:no-fast-pskip

This opens mpv in fullscreen mode with a playlist of two samples. Switch between
samples with "j" and "k" or toggle the original source with "o". Pick the better
sample with "b", the worse with "w" or mark them as equal with "e". New samples
are loaded automatically every time you make a decision. After you've rated all
samples once, any samples that were marked as equal are loaded again. This
process will leave you with the best sample and its settings in the end.

You can adjust the playlist size with the -p (--playlist-size) option if you
want to compare more than two samples.

"Shift+w" does the same thing as "w", but it also removes the sample and its log
file from the file system.

Show and hide the playlist overlay with "`".

Some final notes you might find useful:

- Choose your sample range carefully. It should be representative of the full
  video. 10 seconds or less is fine to check something quickly, but use 60
  seconds or more for fine tuning.

- Increasing gamma with "6" to around 10 to 15 makes differences more obvious.
  (Decrease with "5".) Don't go too high or you'll watch pixels dance that
  nobody else will ever see.

- Try switching between samples while playing. txs should preserve playback
  time, but there is a small delay when mpv seeks.

- You can seek forward and backward by single frames with "." and ",".
'''.strip()


def run():
    argparser = argparse.ArgumentParser(
        prog=__name__,
        formatter_class=MyHelpFormatter,
        description='Generate and compare x264 test encodings with different settings')
    argparser.add_argument('-s', '--source',
                           help='Path to original video')
    argparser.add_argument('-r', '--range', nargs=2, default=['5:00', '10'], metavar=('START', 'DURATION'),
                           help=('Time range in original video; '
                                 'e.g. "10:00 60" means "from 10 minutes to 11 minutes"'))
    argparser.add_argument('-x', '--x264-settings', default='',
                           help=('Colon-separated x264 settings (colons in values must be escaped);'
                                 'subcommands may override these'))
    argparser.add_argument('-d', '--dry-run', action='store_true',
                           help='Only show what would be done with these arguments')
    argparser.add_argument('-o', '--overwrite', action='store_true',
                           help='Overwrite existing files')
    argparser.add_argument('-e', '--estimates-file', default='./estimates', metavar='PATH',
                           help=('Where to store estimates of final size and encoding time;'
                                 'this path is relative to the samples directory'))
    argparser.add_argument('--version', action='version', version=f'{__name__} {__version__}')

    subparsers = argparser.add_subparsers()

    argparser_tutorial = subparsers.add_parser(
        'tutorial',
        help='Show tutorial that explains the basic workflow',
        description='Show tutorial that explains the basic workflow')
    argparser_tutorial.set_defaults(func=lambda args: print(TUTORIAL))

    argparser_samples = subparsers.add_parser(
        'samples',
        help='Generate samples with different settings',
        description='Generate samples with different settings')
    argparser_samples.add_argument('-xs', '--sample-settings', nargs='+', default=[], metavar='SETTINGS',
                                   help='x264 settings to test; values are separated with "/"')
    argparser_samples.set_defaults(func=_samples)

    argparser_compare = subparsers.add_parser(
        'compare',
        formatter_class=MyHelpFormatter,
        help='Compare previously generated samples',
        description='Compare previously generated samples',
        epilog=('key bindings:\n'
                '  j        Play next sample\n'
                '  k        Play previous sample\n'
                '  b        Remove all other samples from playlist\n'
                '           and load more samples\n'
                '  w        Remove current sample from playlist;\n'
                '           load more samples if there is only one left\n'
                '  e        Samples in current playlist are equal; revisit \n'
                '           them after seeing all other samples at least once\n'
                '  shift+w  Delete sample from file system and estimates file\n'
                '  o        Show/Hide original source\n'
                '  `        Show/Hide current playlist\n'
                '\n'
                '  You can change them by putting these lines in ~/.config/mpv/input.conf:\n'
                '    j       script-binding txs/playlist-next\n'
                '    k       script-binding txs/playlist-prev\n'
                '    b       script-binding txs/sample-is-better\n'
                '    w       script-binding txs/sample-is-worse\n'
                '    e       script-binding txs/samples-are-equal\n'
                '    shift+w script-binding txs/sample-is-garbage\n'
                '    o       script-binding txs/toggle-original\n'
                '    `       script-binding txs/toggle-info\n'
                '\n\n'
                'configuration:\n'
                '  The file "script-opts/txs.conf" in mpv\'s user folder (e.g.\n'
                '   ~/.config/mpv/script-opts/txs.conf) can configure these options:\n'
                '     debug=no\n'
                '     playlist_size=5\n'
                '     font_size=8\n'
                '     font_color=FFFFFF\n'
                '     border_size=1.0\n'
                '     border_color=101010\n'
                '     estimates_file=./estimates\n'
        ))

    argparser_compare.add_argument('samples',
                                   help='Directory that contains the samples')
    argparser_compare.add_argument('-p', '--playlist-size', default=None,
                                   help='Maximum number of sample samples to compare')
    argparser_compare.add_argument('-f', '--font-size', default=None,
                                   help='Font size for playlist')
    argparser_compare.add_argument('--debug', action='store_true',
                                   help='Print debugging messages in Lua print')
    argparser_compare.set_defaults(func=_compare)

    argparser_bframes = subparsers.add_parser(
        'bframes',
        help='Generate test encode and show consecutive B-frames percentages',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description=('Generate test encode and show consecutive B-frames percentages\n\n'
                     'The settings given by --x264-settings are used but are optimized\n'
                     'for speed, e.g. crf=51.'))
    argparser_bframes.add_argument('-b', '--bframes', default='16',
                                   help='Maximum number of consecutive B-frames in test encode')
    argparser_bframes.add_argument('-a', '--b-adapt', default='2',
                                   help='Adaptive B-frame decision method')
    argparser_bframes.set_defaults(func=_bframes)

    args = argparser.parse_args()
    if hasattr(args, 'func'):
        args.func(args)
    else:
        argparser.print_help()

def _samples(args):
    base_settings = utils.parse_settings(args.x264_settings)
    sample_settings = utils.generate_sample_settings(*args.sample_settings)
    title = utils.title(args.source)
    samples_dir = os.path.join('.', (f'samples.{title}@{"-".join(args.range)}.' +
                                     ':'.join(utils.sample_keys(sample_settings))))
    if not sample_settings:
        utils.croak('Missing argument: --sample-settings')

    print(f'    Base settings: {utils.settings2str(base_settings, escape=False)}')
    print(f'{len(sample_settings):9d} samples: '
          f'{utils.settings2str(sample_settings, escape=False)}')
    print(f'Samples directory: {samples_dir}')
    if not args.dry_run:
        utils.mkdir(samples_dir)
        # Extract range from original into separate file
        excerpt_path = os.path.join(samples_dir, f'{title}.original@{"-".join(args.range)}.mkv')
        if not os.path.exists(excerpt_path):
            try:
                ffmpeg.encode(args.source, dest=excerpt_path,
                              start=args.range[0], stop=args.range[1],
                              topic=f'  Extracting range {args.range[0]} - {args.range[1]}',
                              create_logfile=False)
            except KeyboardInterrupt:
                print('\n')
                utils.cleanup(excerpt_path)
                utils.croak('Aborted')

    total_secs = ffmpeg.duration(args.source)
    estimates_file = os.path.join(samples_dir, args.estimates_file)
    try:
        for i,diff_settings in enumerate(sample_settings, start=1):
            settings = utils.combine_dicts(base_settings, diff_settings)
            dest = os.path.join(samples_dir,
                                (f'{title}.sample@'
                                 f'{"-".join(args.range)}.'
                                 f'{utils.settings2str(settings, escape=False)}'
                                 f'.mkv'))
            print(f'Sample {i}/{len(sample_settings)}: '
                  f'{utils.settings2str(diff_settings, escape=False)}')
            if not args.dry_run and (args.overwrite or not os.path.exists(dest)):
                start_time = time.monotonic()
                ffmpeg.encode(excerpt_path, dest, settings, topic='  Encoding')
                enc_time = time.monotonic() - start_time
                sample_secs = ffmpeg.duration(dest)
                est_time = enc_time * total_secs / sample_secs
                est_size = os.path.getsize(dest) * total_secs / sample_secs
                utils.update_estimates(estimates_file, diff_settings,
                                       est_time, est_size, settings)
                est = utils.read_estimates(estimates_file)
                key = utils.settings2str(diff_settings, escape=False)
                print(f'  Estimated encoding time: {est[key]["time_str"]}')
                print(f'     Estimated final size: {est[key]["size_str"]}')
            elif os.path.exists(dest):
                print(f'  Already encoded')
                est = utils.read_estimates(estimates_file)
                key = utils.settings2str(diff_settings, escape=False)
                if key in est:
                    print(f'  Estimated encoding time: {est[key]["time_str"]}')
                    print(f'     Estimated final size: {est[key]["size_str"]}')
    except KeyboardInterrupt:
        print('\n')
        utils.cleanup(dest)
        utils.croak('Aborted')
    else:
        if not args.dry_run:
            cmd = [__name__, 'compare', samples_dir]
            print(f'To compare settings visually run:\n{utils.cmd2str(cmd)}')
            if utils.dialog_yesno('Do you want to compare samples now?'):
                utils.compare_samples(samples_dir)


def _compare(args):
    utils.compare_samples(args.samples,
                          debug=args.debug,
                          playlist_size=args.playlist_size,
                          font_size=args.font_size,
                          estimates_file=args.estimates_file)


def _bframes(args):
    title = utils.title(args.source)
    bframes_dir = os.path.join('.', f'bframes:{title}@{"-".join(args.range)}')
    settings = {**utils.parse_settings(args.x264_settings),
                **{# These settings shouldn't change the consecutive bframes
                   # percentages, but they make the test encode faster.
                    'crf': '51',
                    'trellis': '0',
                    'ref': '1',
                    'aq-mode': '0',
                    'partitions': 'none',
                    'weightp': '0',
                    'no-mixed-refs': None,
                    'no-deblock': None,
                    'no-cabac': None,
                    'no-8x8dct': None,
                    'no-scenecut': None},
                **{'bframes': args.bframes, 'b-adapt': args.b_adapt}}
    dest = os.path.join(bframes_dir,
                        (f'{title}.bframes@{"-".join(args.range)}.'
                         f'{utils.settings2str(settings, escape=False)}'
                         f'.mkv'))
    print(f'Finding consecutive B-frames with these settings:')
    print(utils.settings2str(settings, escape=False))
    try:
        if not args.dry_run:
            utils.mkdir(bframes_dir)
            if args.overwrite or not os.path.exists(utils.logfile(dest)):
                ffmpeg.encode(args.source, dest, settings,
                              start=args.range[0], stop=args.range[1])
    except KeyboardInterrupt:
        print('\n')
        utils.cleanup(dest)
        utils.croak('Aborted')
    else:
        if not args.dry_run:
            bframes = ffmpeg.bframes(utils.logfile(dest))
            lines = [' '.join(f' {i:2d}  ' for i in range(17))]
            values = []
            for perc in bframes:
                values.append(f'{perc:4.1f}%')
            lines.append(' '.join(values))
            lines.append(utils.wrap('For each possible number of consecutive B-frames, '
                                    'show how many frames (in percent) are in such a '
                                    'sequence of B-frames.', width=100))
            for line in lines:
                print(line)
            with open(os.path.join(bframes_dir, 'bframes'), 'w') as f:
                for line in lines:
                    print(line, file=f)
